x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/async/AsyncResultsService.java

```diff
@@ -120,16 +120,11 @@ private void getSearchResponseFromTask(
     ) {
         try {
             final Task task = store.getTaskAndCheckAuthentication(taskManager, searchId, asyncTaskClass);
-            if (task == null) {
+            if (task == null || task.isCancelled()) {
                 getSearchResponseFromIndex(searchId, request, nowInMillis, listener);
                 return;
             }
 
-            if (task.isCancelled()) {
-                listener.onFailure(new ResourceNotFoundException(searchId.getEncoded()));
-                return;
-            }
```

Comment:
Based on my analysis and testing (both manual and automated), this fail-fast if block is not needed and actually undesirable.

When a long-running search is cancelled via the `_tasks/_cancel` API but before the task and search fully complete, the GET `_async_search/:id` endpoint throws this "resource_not_found_exception". But then when the task is unregistered and the search is fully finished, the GET `_async_search/:id` endpoint no longer throws that exception, but instead returns the expected output showing that correct error (that the search was cancelled).

With this change, that no longer happens. The GET `_async_search/:id` endpoint returns useful information (like the `GET `_async_search/status/:id` endpoint` already does) after the cancellation but before the search has fully shutdown, as shown below.

<details><summary><code>GET _async_search/:id</code> after task is cancelled but it is not yet unregistered and the search has not fully shutdown</summary>

```
{
  "id": "FkswMWdMc2VEUVJxTXNMVGtnaUliU3caMl90V1VTZHBUd1d4WEx5dDhwMlhCdzoxMjg=",
  "is_partial": true,
  "is_running": true,
  "start_time_in_millis": 1686667932790,
  "expiration_time_in_millis": 1687099932790,
  "response": {
    "took": 131,
    "timed_out": false,
    "terminated_early": false,
    "num_reduce_phases": 0,
    "_shards": {
      "total": 3,
      "successful": 0,
      "skipped": 0,
      "failed": 0
    },
    "_clusters": {
      "total": 3,
      "successful": 0,
      "skipped": 0
    },
    "hits": {
      "total": {
        "value": 0,
        "relation": "gte"
      },
      "max_score": "",
      "hits": []
    }
  }
}
```

</details>

<details><summary><code>GET _async_search/:id</code> after task is unregistered and the search has fully shutdown</summary>

```js
{
  "id": "FkswMWdMc2VEUVJxTXNMVGtnaUliU3caMl90V1VTZHBUd1d4WEx5dDhwMlhCdzoxMjg=",
  "is_partial": true,
  "is_running": false,
  "start_time_in_millis": 1686667932790,
  "expiration_time_in_millis": 1687099932790,
  "response": {
    "took": 36810,
    "timed_out": false,
    "terminated_early": false,
    "num_reduce_phases": 0,
    "_shards": {
      "total": 3,
      "successful": 0,
      "skipped": 0,
      "failed": 3,
      "failures": [
        {
          "shard": 0,
          "index": "blogs",
          "node": "2_tWUSdpTwWxXLyt8p2XBw",
          "reason": {
            "type": "task_cancelled_exception",
            "reason": "task cancelled [by user request]"
          }
        }
      ]
    },
    "_clusters": {
      "total": 3,
      "successful": 0,
      "skipped": 0
    },
    "hits": {
      "total": {
        "value": 0,
        "relation": "gte"
      },
      "max_score": "",
      "hits": []
    }
  },
  "error": {
    "type": "status_exception",
    "reason": "error while executing search",
    "caused_by": {
      "type": "search_phase_execution_exception",
      "reason": "all shards failed",
      "phase": "query",
      "grouped": true,
      "failed_shards": [
        {
          "shard": 0,
          "index": "remote1:blogs",
          "node": "tsYK9MXVRcezptK7HYwLLA",
          "reason": {
            "type": "task_cancelled_exception",
            "reason": "task cancelled [by user request]"
          }
        }
      ],
      "caused_by": {
        "type": "task_cancelled_exception",
        "reason": "task cancelled [by user request]"
      }
    }
  }
}
```

</details>

This behavior now matches how the `GET _async_search/status/:id` endpoint behaves under these conditions as well (not shown).

Reply:
The other scenario to consider for this code change is when the search is deleted via `DELETE _async_search`. In that throwing `ResourceNotFoundException` is correct. However, in this case, there is also no need to have this short-circuit block, as the call below it to [sendFinalResponse](https://github.com/elastic/elasticsearch/blob/b2504a9f87cb854ce35214b23bda41d600cca2b2/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/async/AsyncResultsService.java#L133) will also throw `ResourceNotFoundException`.

Reply:
thanks for sharing all the context!